In [1]:
import numpy as np

# Points as given in the problem
points = np.array([
    [2, 10],#p1
    [2, 5],#p2
    [8, 4],#p3
    [5, 8],#p4
    [7, 5],#p5
    [6, 4],#p6
    [1, 2],#p7
    [4, 9],#p8
])
# Initial centroids
m1 = points[0]  # P1
m2 = points[7]  # P8
m3 = points[4]  # P5

# Function to calculate Euclidean distance
def euclidean_distance(point, centroid):
    return np.sqrt(np.sum((point - centroid) ** 2))

# Perform one iteration of k-means
def k_means_iteration(points, m1, m2, m3):
    cluster1, cluster2, cluster3 = [], [], []
    for point in points:
        # Calculate distances to centroids
        dist_to_m1 = euclidean_distance(point, m1)
        dist_to_m2 = euclidean_distance(point, m2)
        dist_to_m3 = euclidean_distance(point, m3)

        # Assign point to the nearest cluster
        if dist_to_m1 < dist_to_m2 and dist_to_m1 < dist_to_m3:
            cluster1.append(point)
        elif dist_to_m2 < dist_to_m1 and dist_to_m2 < dist_to_m3:
            cluster2.append(point)
        else:
            cluster3.append(point)

    # Update centroids
    new_m1 = np.mean(cluster1, axis=0) if cluster1 else m1
    new_m2 = np.mean(cluster2, axis=0) if cluster2 else m2
    new_m3 = np.mean(cluster3, axis=0) if cluster3 else m3
    return np.array(cluster1), np.array(cluster2), np.array(cluster3), new_m1, new_m2, new_m3

# Run one iteration
cluster1, cluster2, cluster3, new_m1, new_m2, new_m3 = k_means_iteration(points, m1, m2, m3)

# Results
print("Cluster 1 (around m1):", cluster1)
print("Cluster 2 (around m2):", cluster2)
print("Cluster 3 (around m3):", cluster3)
print("Updated centroid m1:", new_m1)
print("Updated centroid m2:", new_m2)
print("Updated centroid m3:", new_m3)




Cluster 1 (around m1): [[ 2 10]]
Cluster 2 (around m2): [[2 5]
 [5 8]
 [4 9]]
Cluster 3 (around m3): [[8 4]
 [7 5]
 [6 4]
 [1 2]]
Updated centroid m1: [ 2. 10.]
Updated centroid m2: [3.66666667 7.33333333]
Updated centroid m3: [5.5  3.75]


#QUESTIONS AND ANSWERS
#1.Which cluster does P6[6,4] belong to?
Ans=P6 belongs to cluster cluster 3
#2. What is the population of a cluster around m3?
Ans=4 (i.e p3,p5,p6,p7 three data points are present in cluster m2)
#3. What is the updated value of m1 and m2?
Ans=From above resuls,
Updated centroid m1: [ 2. 10.]
Updated centroid m2: [3.66666667 7.33333333]
Updated centroid m3: [5.5  3.75]